samples/echo/src/main/cpp/AudioEffect.cpp

```diff
@@ -0,0 +1,277 @@
+          int32_t sampleRate, int32_t channelCount, oboe::AudioFormat format){
+  if (busy_) {
+    LOGW("filtering in progress, filter configuration is IGNORED");
+    return;
+  }
+  bgAudio_ = std::move(samples);
+  bgAudioSampleCount_ = sampleCount;
+  sampleRate_ = sampleRate;
+  format_ = format;
+  channelCount_ = channelCount;
+
+  curPosition_ = 0;
+}
+
+/**
+ * Adding audio processing into the live audio
+ * @param liveAudio is recorded audio
+ * @param samplesPerFrame is same as channelCount.
+ * @param numFrames represents frames pointed by liveAudio
+ *        total samples = numFrames * samplesPerFrame
+ */
+void AudioMixer::process(int16_t *liveAudio, int32_t channelCount,
+                          int32_t numFrames) {
+  assert(bgAudio_ && liveAudio);
+  if (numFrames > bgAudioSampleCount_ || channelCount != channelCount_ ||
+      bgMixFactorInt_ == 0) {
+    return;
+  }
+
+  busy_ = true;
+  int32_t curSample;
+  for (int i = 0; i < (numFrames * channelCount); i++) {
+    curSample = liveAudio[i];
+    curSample = curSample * fgMixFactorInt_ +
+                bgAudio_[curPosition_] * bgMixFactorInt_;
+    curSample /= kFloatToIntMapFactor;
+
+    curSample = (curSample > SHRT_MAX ? SHRT_MAX : curSample);
+    liveAudio[i] = (int16_t)(curSample < SHRT_MIN ? SHRT_MIN : curSample);
+    curPosition_ = (curPosition_ + 1 ) % bgAudioSampleCount_;
+  }
+  busy_ = false;
+}
+
+/**
+ * query for audio format supportability
+ */
+bool AudioMixer::AudioFormatSupported(int32_t frequency,
+                  int32_t channels, oboe::AudioFormat format) const {
+  return (frequency == sampleRate_ &&
+          channels == channelCount_ &&
+          format == format_);
+}
+
+/**
+ * Constructor for AudioDelay
+ * @param sampleRate
+ * @param channelCount
+ * @param format
+ * @param delay
+ * @param decay
+ */
+AudioDelay::AudioDelay(int32_t sampleRate,
+                       int32_t channelCount,
+                       oboe::AudioFormat format,
+                       float  delay,
+                       float  decay) :
+    AudioFormat(sampleRate, channelCount, format),
+    delay_(delay), decay_(decay) {
+
+  feedbackFactor_ = static_cast<int32_t>(decay_ * kFloatToIntMapFactor);
+  liveAudioFactor_ = kFloatToIntMapFactor - feedbackFactor_;
+  allocateBuffer();
+}
+
+/**
+ * Destructor
+ */
+AudioDelay::~AudioDelay() {
+  if(buffer_) delete static_cast<uint8_t*>(buffer_);
+}
+
+/**
+ * Configure for delay time ( in miliseconds ). It is possible to dynamically
+ * adjust the value
+ * @param delay in seconds
+ * @return true if delay time is set successfully
+ */
+bool AudioDelay::setDelay(float delay) {
+  float delta = delay - delay_;
+  if ( delta > -0.022f && delta < 0.022f) {
+    return true;
+  }
+
+  std::lock_guard<std::mutex> lock(lock_);
+
+  if(buffer_) {
+    delete static_cast<uint8_t*>(buffer_);
+    buffer_ = nullptr;
+  }
+
+  delay_  = delay;
+  allocateBuffer();
+  return buffer_ != nullptr;
```

Comment:
I don't see how `buffer_` could be `nullptr` here. The current implementation of `allocateBuffer` will crash the app if any problem occurs.